src/app_service/service/provider/service.nim

```diff
@@ -108,196 +45,18 @@ proc newService*(dappPermissionsService: dapp_permissions_service.ServiceInterfa
-      let success: bool = not jsonRpcResult.hasKey("error")
-      let errorMessage = if success: "" else: jsonRpcResult["error"]{"message"}.getStr()
-      let response = if success: jsonRpcResult["result"].getStr() else: ""
-
-      return $ %* {
-        "type": ResponseTypes.Web3SendAsyncCallback,
-        "messageId": data.messageId,
-        "error": errorMessage,
-        "result": {
-          "jsonrpc": "2.0",
-          "id": if data.payload.id == nil: newJNull() else: data.payload.id,
-          "result": if (success): response else: ""
-        }
-      }
-
-    except Exception as e:
-        error "Error signing message", msg = e.msg
-        return $ %* {
-          "type": ResponseTypes.Web3SendAsyncCallback,
-          "messageId": data.messageId,
-          "error": {
-            "code": 4100,
-            "message": e.msg
-          }
-        }
-
-
-
-  if ACC_METHODS.contains(data.payload.rpcMethod):
-      let dappAddress = self.settingsService.getDappsAddress()
-      return $ %* {
-        "type": ResponseTypes.Web3SendAsyncCallback,
-        "messageId": data.messageId,
-        "result": {
-          "jsonrpc": "2.0",
-          "id": data.payload.id,
-          "result": if data.payload.rpcMethod == "eth_coinbase": newJString(dappAddress) else: %*[dappAddress]
-          }
-        }
-
-  let rpcResult = callRPC(data.request)
-
-  return $ %* {
-    "type": ResponseTypes.Web3SendAsyncCallback,
-    "messageId": data.messageId,
-    "error": (if rpcResult == "": newJString("web3-response-error") else: newJNull()),
-    "result": rpcResult.parseJson
-  }
-
-proc process(self: Service, data: APIRequest): string =
-  var value:JsonNode = case data.permission
-  of Permission.Web3: %* [self.settingsService.getDappsAddress()]
-  of Permission.ContactCode: %* self.settingsService.getPubKey()
-  of Permission.Unknown: newJNull()
-
-  let isAllowed = data.isAllowed and data.permission != Permission.Unknown
-
-  info "API request received", host=data.hostname, value=data.permission, isAllowed
-
-  if isAllowed: 
-    discard self.dappPermissionsService.addPermission(data.hostname, data.permission)
-
-  return $ %* {
-    "type": ResponseTypes.APIResponse,
-    "isAllowed": isAllowed,
-    "permission": data.permission,
-    "messageId": data.messageId,
-    "data": value
-  }
-
-method postMessage*(self: Service, message: string): string =
-  case message.requestType():
-  of RequestTypes.Web3SendAsyncReadOnly: self.process(message.toWeb3SendAsyncReadOnly())
-  of RequestTypes.HistoryStateChanged: """{"type":"TODO-IMPLEMENT-THIS"}""" ############# TODO:
-  of RequestTypes.APIRequest: self.process(message.toAPIRequest())
-  else:  """{"type":"TODO-IMPLEMENT-THIS"}""" ##################### TODO:
 
+method postMessage*(self: Service, requestType: string, message: JsonNode): string =
+  try:
+    if requestType == "web3-send-async-read-only":
+      if message.hasKey("payload"):
+        if message["payload"].hasKey("password"):
```

Comment:
nit: rather than 3 level of if you can use: `&&` it doesn't change much so adding as nit and let you pick the one you prefer